edb/lang/edgeql/compiler/context.py

```diff
@@ -112,8 +112,16 @@ class ContextLevel(compiler.ContextLevel):
     aliased_views: typing.Dict[str, s_nodes.Node]
     """A dictionary of views aliased in a statement body."""
 
-    view_class_map: typing.Dict[s_nodes.Node, s_nodes.Node]  # noqa
-    """Object mapping (used by schema-level views)."""
+    schema_view_cache: typing.Dict[s_nodes.Node, s_nodes.Node]  # noqa
+    """Type cache used by schema-level views."""
+
+    expr_view_cache: typing.Dict[typing.Tuple[qlast.Base, str],
+                                 irast.Set]  # noqa
+    """Type cache used by expression-level views."""
+
+    shape_type_cache: typing.Dict[typing.Tuple[qlast.ShapeElement],
```

Comment:
Is key a tuple of single `qlast.ShapeElement` element?

Reply:
Good catch. It's a tuple of many.